src/keygen/participant.rs

```diff
@@ -317,15 +314,23 @@ impl KeygenParticipant {
         rng: &mut R,
         broadcast_message: BroadcastOutput,
     ) -> Result<ProcessOutcome<<Self as ProtocolParticipant>::Output>> {
-        info!("Handling round one keygen message.");
+        let message = broadcast_message.into_message(BroadcastTag::KeyGenR1CommitHash)?;
 
-        // XXX should we have a check that we haven't recieved a round one
-        // message _after_ round one is complete? Likewise for all other rounds.
+        if self
```

Comment:
I think we should factor out this code into a reusable function:
- We basically have the same code three times in this file, and we will end up with similar code on other files.
- With the current implementation the caller gets back a `Ok(ProcessOutcome::Incomplete)`. So the caller cannot programmatically check if a duplicate message was found. This info is only available via the logs. 

My suggestions is the following function:
```rust
    fn check_for_duplicate_msg<T: TypeTag>(&self, from: ParticipantIdentifier) -> Result<()> {
        if self.local_storage().contains::<T>(from) {
            warn!(
                "Received duplicate message from {:?} of type {}",
                from,
                std::any::type_name::<T>()
            );
            return Err(InternalError::DuplicateMessage);
        } else {
            Ok(())
        }
    }
```
I added a new `InternalError` here called `InternalError::DuplicateMessage`. So the user can match on this if they choose.

This `check_for_duplicate_msg` function should be part of the trait definition of `trait InnerProtocolParticipant`. This way any `Participants` that impls `InnerProtocolParticipant` will automatically have access to it.

Then we can simply change this check to:
```rust
self.check_for_duplicate_msg::<storage::Commit>(message.from())?;
```

LMK if you have questions or concerns with this approach!

Reply:
Good idea, I did that.